core/src/main/java/org/sonarsource/sonarlint/core/AutoBinding.java

```diff
@@ -0,0 +1,408 @@
+        cluesAndConnections.add(new BindingClueWithConnections(bindingClue, connectionsIds));
+      }
+    }
+    return cluesAndConnections;
+  }
+
+  private static class BindingClueWithConnections {
+    private final BindingClue bindingClue;
+    private final Set<String> connectionIds;
+
+    private BindingClueWithConnections(BindingClue bindingClue, Set<String> connectionIds) {
+      this.bindingClue = bindingClue;
+      this.connectionIds = connectionIds;
+    }
+  }
+
+  private List<BindingClue> collectBindingClues(String checkedConfigScopeId) {
+    FindFileByNamesInScopeResponse response;
+    try {
+      response = client.findFileByNamesInScope(new FindFileByNamesInScopeParams(checkedConfigScopeId, List.of(SONAR_SCANNER_CONFIG_FILENAME, AUTOSCAN_CONFIG_FILENAME))).get(1,
+        TimeUnit.MINUTES);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new IllegalStateException(e);
+    } catch (ExecutionException e) {
+      LOG.error("Unable to search scanner clues", e.getCause());
+      return List.of();
+    } catch (TimeoutException e) {
+      LOG.error("Unable to search scanner clues in time", e);
+      return List.of();
+    }
+
+    List<BindingClue> bindingClues = new ArrayList<>();
+    for (FoundFile foundFile : response.getFoundFiles()) {
+      var scannerProps = extractScannerProperties(foundFile);
+      if (scannerProps == null) {
+        continue;
+      }
+      var bindingClue = computeBindingClue(foundFile.getFilename(), scannerProps);
+      if (bindingClue != null) {
+        bindingClues.add(bindingClue);
+      }
+    }
+    return bindingClues;
+  }
+
+  private Set<String> matchConnections(BindingClue bindingClue) {
+    if (bindingClue instanceof SonarQubeBindingClue) {
+      var serverUrl = ((SonarQubeBindingClue) bindingClue).serverUrl;
+      return connectionReferential.getConnectionsById().values().stream()
+        .filter(SonarQubeConnectionConfiguration.class::isInstance)
+        .map(SonarQubeConnectionConfiguration.class::cast)
+        .filter(c -> isSameServerUrl(c.getServerUrl(), serverUrl))
+        .map(AbstractConnectionConfiguration::getConnectionId)
+        .collect(toSet());
+    }
+    if (bindingClue instanceof SonarCloudBindingClue) {
+      var organization = ((SonarCloudBindingClue) bindingClue).organization;
+      return connectionReferential.getConnectionsById().values().stream()
+        .filter(SonarCloudConnectionConfiguration.class::isInstance)
+        .map(SonarCloudConnectionConfiguration.class::cast)
+        .filter(c -> organization == null || Objects.equals(organization, c.getOrganization()))
+        .map(AbstractConnectionConfiguration::getConnectionId)
+        .collect(toSet());
+    }
+    return connectionReferential.getConnectionsById().keySet();
+  }
+
+  private static boolean isSameServerUrl(String left, String right) {
+    return Objects.equals(removeEnd(left, "/"), removeEnd(right, "/"));
```

Comment:
I would expect to find this in the SonarQubeConnectionConfiguration directly.
But then I realized this class is the DTO that we use in the RPC method. So my suggestion is to introduce a separate `Connection` type, in which we could add e.g. `Connection#hasSameHost(Connection)`
I don't think we should rely on the DTO for all our operations in the backend

Reply:
Created DTOs